core/datasource/common/time.go

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) 2023 Gobalsky Labs Limited
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE.VEGA file and at https://www.mariadb.com/bsl11.
+//
+// Change Date: 18 months from the later of the date of the first publicly
+// available Distribution of this version of the repository, and 25 June 2022.
+//
+// On the date above, in accordance with the Business Source License, use
+// of this software will be governed by version 3 or later of the GNU General
+// Public License.
+
+//lint:file-ignore ST1003 Ignore underscores in names, this is straigh copied from the proto package to ease introducing the domain types
+
+package common
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"code.vegaprotocol.io/vega/libs/ptr"
+	datapb "code.vegaprotocol.io/vega/protos/vega/data/v1"
+)
+
+type InternalTimeTrigger struct {
+	// This is optional to reflect the proto, but it will always be set by the governance
+	Initial     *time.Time
+	Every       int64
+	NextTrigger *time.Time
+}
+
+func (i InternalTimeTrigger) String() string {
+	return fmt.Sprintf(
+		"initial(%v) every(%d) nextTrigger(%v)",
+		i.Initial,
+		i.Every,
+		i.NextTrigger,
+	)
+}
+
+func (i InternalTimeTrigger) IntoProto() *datapb.InternalTimeTrigger {
+	var initial *int64
+	if i.Initial != nil {
+		initial = ptr.From(i.Initial.Unix())
+	}
+
+	return &datapb.InternalTimeTrigger{
+		Initial: initial,
+		Every:   i.Every,
+	}
+}
+
+func (i InternalTimeTrigger) DeepClone() *InternalTimeTrigger {
+	var initial *time.Time
+	if i.Initial != nil {
+		initial = *ptr.From(i.Initial)
+	}
+
+	var nextTrigger *time.Time
+	if i.NextTrigger != nil {
+		nextTrigger = *ptr.From(i.NextTrigger)
+	}
+
+	return &InternalTimeTrigger{
+		Initial:     initial,
+		Every:       i.Every,
+		NextTrigger: nextTrigger,
+	}
+}
+
+func (i InternalTimeTrigger) IsTriggered(timeNow time.Time) bool {
+	if i.NextTrigger.Before(timeNow) {
+		*i.NextTrigger = i.NextTrigger.Add(time.Duration(i.Every) * time.Second)
+		return true
+	}
+
+	return false
+}
+
+func (i InternalTimeTrigger) SetNextTrigger(timeNow time.Time) {
```

Comment:
this should probably ne private, and called only on the constructor / FromProto method.

Reply:
In that case there will be a getter for this field, otherwise it cannot be tested - is this what you mean ?

Reply:
well, IsTriggered is the way to test it isn't it?

Reply:
As testing the boolean result yes, i meant the setting of the actual value of this trigger
Do we want to test that as well ?

Reply:
hum I don't think it matters really, at the end of the day you want to test the overall behaviour of the type, not the actual value it contains, which is meant to be abstracted away I suppose.

Reply:
All right, I'll remove it then

Reply:
This has been corrected now, as well as the tests.